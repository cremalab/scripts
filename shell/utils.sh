#!/bin/bash
function print {
  echo ""
  echo "-----------------------------------------------------"
  echo $1
  echo "-----------------------------------------------------"
}