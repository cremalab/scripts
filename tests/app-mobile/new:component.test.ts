import spawn from "cross-spawn"
import { cleanup } from "../utils/cleanup"
import { generateNewComponent } from "../utils/generateNewComponent"
import { navigateToConsumer } from "../utils/navigateToConsumer"
import { printFileContents } from "../utils/printFileContents"
import { printTreeOfNew } from "../utils/printTreeOfNew"

beforeAll(navigateToConsumer("app-mobile"))

describe("new:component", () => {
  // Component imports are added here and must be removed after the tests
  afterAll(() => spawn.sync("git", ["checkout", "./src/stories/index.tsx"], {
    stdio: "inherit",
  }))

  describe("without example code", () => {
    beforeAll(() =>
      generateNewComponent({ name: "A", withExample: false, withState: false }),
    )
    
    afterAll(() => cleanup("./src/components/A"))

    it("generates directory with expected files", async () => {
      const tree = printTreeOfNew("components", "A")
      expect(tree).toMatchSnapshot()
    })

    it("generates directory with expected `index.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/A/index.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `README.md` contents", async () => {
      const fileContents = await printFileContents("./src/components/A/README.md")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `stories.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/A/stories.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `test.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/A/test.tsx")
      expect(fileContents).toMatchSnapshot()
    })
  })

  describe("with example but without state", () => {
    beforeAll(() =>
      generateNewComponent({ name: "B", withExample: true, withState: false }),
    )

    afterAll(() => cleanup("./src/components/B"))
    
    it("generates component with expected files", async () => {
      const tree = printTreeOfNew("components", "B")
      expect(tree).toMatchSnapshot()
    })

    it("generates directory with expected `index.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/B/index.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `README.md` contents", async () => {
      const fileContents = await printFileContents("./src/components/B/README.md")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `stories.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/B/stories.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `test.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/B/test.tsx")
      expect(fileContents).toMatchSnapshot()
    })
  })

  describe("with example and with state", () => {
    beforeAll(() =>
      generateNewComponent({ name: "C", withExample: true, withState: true }),
    )
    
    afterAll(() => cleanup("./src/components/C"))

    it("generates component with expected files", async () => {
      const tree = printTreeOfNew("components", "C")
      expect(tree).toMatchSnapshot()
    })

    it("generates directory with expected `index.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/C/index.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `README.md` contents", async () => {
      const fileContents = await printFileContents("./src/components/C/README.md")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `stories.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/C/stories.tsx")
      expect(fileContents).toMatchSnapshot()
    })

    it("generates directory with expected `test.tsx` contents", async () => {
      const fileContents = await printFileContents("./src/components/C/test.tsx")
      expect(fileContents).toMatchSnapshot()
    })
  })
})
