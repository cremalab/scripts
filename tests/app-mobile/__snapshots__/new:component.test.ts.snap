// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new:component with example code and with state generates expected C.stories.tsx contents 1`] = `
"import { storiesOf } from \\"@storybook/react-native\\"
import React from \\"react\\"
import { C } from \\"./C\\"

storiesOf(\\"C\\", module).add(\\"example\\", () => <C />)
"
`;

exports[`new:component with example code and with state generates expected C.test.tsx contents 1`] = `
"import { C } from \\"./C\\"
import { render, fireEvent } from \\"@testing-library/react-native\\"
import React from \\"react\\"

describe(\\"C\\", () => {
  it(\\"has correct message before clicking\\", () => {
    // Arrange
    const message = \\"You clicked 0 times\\"

    // Act
    const { getByText } = render(<C />)
    const received = getByText(message)

    // Assert
    expect(received).toBeDefined()
  })

  it(\\"has correct message after clicking\\", () => {
    // Arrange
    const name = \\"Click Me\\"
    const message = \\"You clicked 1 times\\"

    // Act
    const { getByText } = render(<C />)
    const button = getByText(name)
    fireEvent.press(button)
    const received = getByText(message)

    // Assert
    expect(received).toBeDefined()
  })
})
"
`;

exports[`new:component with example code and with state generates expected C.tsx contents 1`] = `
"import React, { useState } from \\"react\\"
import { Button, StyleSheet, Text, View } from \\"react-native\\"

type Props = Readonly<{}>

export function C(_: Props) {
  const [count, setCount] = useState(0)
  return (
    <View style={styles.container}>
      <Text>You clicked {count} times</Text>
      <Button onPress={() => setCount(count + 1)} title=\\"Click Me\\" />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {},
})
"
`;

exports[`new:component with example code and with state generates expected README.md contents 1`] = `
"# \`<C />\`

DESCRIPTION_HERE

## Directory Structure

- \`stories.tsx\`: Component playground (\`npm run test:playground\`)
- \`test.tsx\`: Component tests (\`npm run test:unit\`)
- \`index.tsx\`: Component code
- \`README.md\`: Component documentation (hey, that's me!)

## Example

![C](../../../.loki/reference/chrome_C_example.png)

> Image location: [\`.loki/reference/chrome_C_example.png\`](../../../.loki/reference/chrome_C_example.png)
> 
> To regenerate: 
> 1. \`npm run test:playground\` (skip if running)
> 1. \`npm run test:visual:update -- --storiesFilter=\\"^C example\\\\$\\"\`
"
`;

exports[`new:component with example code and with state generates expected index.ts contents 1`] = `
"export { C } from \\"./C\\"
"
`;

exports[`new:component with example code but without state generates expected B.stories.tsx contents 1`] = `
"import { storiesOf } from \\"@storybook/react-native\\"
import React from \\"react\\"
import { B } from \\"./B\\"

storiesOf(\\"B\\", module).add(\\"example\\", () => <B />)
"
`;

exports[`new:component with example code but without state generates expected B.test.tsx contents 1`] = `
"import { B } from \\"./B\\"
import { render } from \\"@testing-library/react-native\\"
import React from \\"react\\"

describe(\\"B\\", () => {
  it(\\"name prop is rendered\\", () => {
    // Arrange
    const name = \\"B\\"

    // Act
    const { getByText } = render(<B />)
    const received = getByText(name)

    // Assert
    expect(received).toBeDefined()
  })
})
"
`;

exports[`new:component with example code but without state generates expected B.tsx contents 1`] = `
"import React from \\"react\\"
import { StyleSheet, Text, View } from \\"react-native\\"

type Props = Readonly<{}>

export function B(_: Props) {
  return (
    <View style={styles.container}>
      <Text>B</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {},
})
"
`;

exports[`new:component with example code but without state generates expected README.md contents 1`] = `
"# \`<B />\`

DESCRIPTION_HERE

## Directory Structure

- \`stories.tsx\`: Component playground (\`npm run test:playground\`)
- \`test.tsx\`: Component tests (\`npm run test:unit\`)
- \`index.tsx\`: Component code
- \`README.md\`: Component documentation (hey, that's me!)

## Example

![B](../../../.loki/reference/chrome_B_example.png)

> Image location: [\`.loki/reference/chrome_B_example.png\`](../../../.loki/reference/chrome_B_example.png)
> 
> To regenerate: 
> 1. \`npm run test:playground\` (skip if running)
> 1. \`npm run test:visual:update -- --storiesFilter=\\"^B example\\\\$\\"\`
"
`;

exports[`new:component with example code but without state generates expected index.ts contents 1`] = `
"export { B } from \\"./B\\"
"
`;

exports[`new:component without example code generates expected A.stories.tsx contents 1`] = `
"import { storiesOf } from \\"@storybook/react-native\\"
import React from \\"react\\"
import { A } from \\"./A\\"

storiesOf(\\"A\\", module).add(\\"example\\", () => <A />)
"
`;

exports[`new:component without example code generates expected A.test.tsx contents 1`] = `
"import { A } from \\"./A\\"
import React from \\"react\\"

describe(\\"A\\", () => {
  it(\\"is defined\\", () => expect(<A />).toBeDefined())

  it.todo(\`add meaningful tests üëç\`)
})
"
`;

exports[`new:component without example code generates expected A.tsx contents 1`] = `
"import React from \\"react\\"

export function A() {
  return <></>
}
"
`;

exports[`new:component without example code generates expected README.md contents 1`] = `
"# \`<A />\`

DESCRIPTION_HERE

## Directory Structure

- \`stories.tsx\`: Component playground (\`npm run test:playground\`)
- \`test.tsx\`: Component tests (\`npm run test:unit\`)
- \`index.tsx\`: Component code
- \`README.md\`: Component documentation (hey, that's me!)

## Example

![A](../../../.loki/reference/chrome_A_example.png)

> Image location: [\`.loki/reference/chrome_A_example.png\`](../../../.loki/reference/chrome_A_example.png)
> 
> To regenerate: 
> 1. \`npm run test:playground\` (skip if running)
> 1. \`npm run test:visual:update -- --storiesFilter=\\"^A example\\\\$\\"\`
"
`;

exports[`new:component without example code generates expected index.ts contents 1`] = `
"export { A } from \\"./A\\"
"
`;
